logstash-core/lib/logstash/settings.rb

```diff
@@ -364,17 +372,78 @@ def initialize(name, default=nil, strict=true)
     end
 
     class WritableDirectory < Setting
-      def initialize(name, default=nil, strict=true)
-        super(name, ::String, default, strict) do |path|
-          if ::File.directory?(path) && ::File.writable?(path)
-            true
-          else
-            raise ::ArgumentError.new("Path \"#{path}\" is not a directory or not writable.")
+      def initialize(name, default=nil, strict=false)
+        super(name, ::String, default, strict)
+      end
+      
+      def validate(path)
+        super(path)
+
+        if ::File.directory?(path)
+          if !::File.writable?(path)
+            raise ::ArgumentError.new("Path \"#{path}\" must be a writable directory. It is not writable.")
+          end
+        elsif ::File.symlink?(path)
+          # TODO(sissel): I'm OK if we relax this restriction. My experience
+          # is that it's usually easier and safer to just reject symlinks.
+          raise ::ArgumentError.new("Path \"#{path}\" must be a writable directory. It cannot be a symlink.")
+        elsif ::File.exist?(path)
+          raise ::ArgumentError.new("Path \"#{path}\" must be a writable directory. It is not a directory.")
+        else
+          parent = ::File.dirname(path)
+          if !::File.writable?(parent)
+            raise ::ArgumentError.new("Path \"#{path}\" does not exist and I cannot create it because the parent path \"#{parent}\" is not writable.")
           end
         end
+
+        # If we get here, the directory exists and is writable.
+        true
+      end
+
+      def value
+        super.tap do |path|
+          if !::File.directory?(path)
+            # Create the directory if it doesn't exist.
```

Comment:
Two use cases. First, for the PQ that's coming -- the default path si probably going to be `{path.data}/queue/...` and it'd be nice not to have to have all these empty directories. Second, for users who just want to specify `{path.data}` and that would imply the queue path (or similar) would be relative to `{path.data}` and would be nice if they were automatically created if possible. My hope is to avoid having our docs say `Run these 7 mkdir commands before running logstash.`